README.md

```diff
@@ -1,16 +1,715 @@
+            "subject": "DNSLink for example.com",
+            "body": "Hello Bob!"
+          }
+        ]
+      },
+      "log-as-done": {
+        "with": "https://example.com/report"
+        "do": "crud/update"
+        "inputs": {
+          "from": "mailto://alice@exmaple.com",
+          "to": ["bob@exmaple.com"],
+          "event": "email-notification",
+          "value": {"ucan/promise": ["/", "notify-bob"]} // Pipelined promise
+        }
+      }
+    }
+  },
+  "sig": {"/": {"bytes:": "5vNn4--uTeGk_vayyPuNTYJ71Yr2nWkc6AkTv1QPWSgetpsu8SHegWoDakPVTdxkWb6nhVKAz6JdpgnjABppC7"}}
+}
+```
+
+# 4 Receipt
+
+An invocation receipt is an attestation of the output of an invocation. A receipt MUST be signed by the executor (the `aud` of the associated UCAN).
+
+**NB: a Receipt this does not guarantee correctness of the result!** The statement's veracity MUST be only understood as an attestation from the executor.
+
+Receipts don't have their own version field. Receipts MUST use the same version as the invocation that they contain.
+
+## 4.1 Fields
+
+| Field  | Type            | Description                                                           | Required | Default |
+|--------|-----------------|-----------------------------------------------------------------------|----------|---------|
+| `inv`  | `&Invocation`   | CID of the Invocation that generated this response                    | Yes      |         |
+| `out`  | `{String: Any}` | The results of each call, the task's label. MAY contain sub-receipts. | Yes      |         |
+| `meta` | `Any`           | Non-normative extended fields                                         | No       | `null`  |
+
+### 4.1.1 Invocation
+
+The `inv` field MUST include a link to the Invocation that the Receipt is for.
+
+### 4.1.2 Output
+
+The `out` field MUST contain the output of steps of the call graph, indexed by the task name inside the invocation. If the invocation is the implicit `"*"`, then the base64 hash of the concatenation of the URI, Ability and extensional fields MUST be used.
+
+The `out` field MAY omit any tasks that have not yet completed, or results which are not public. An `Invocation` may be associated to zero or more `Receipts`.
+
+A `Result` MAY include recursive `Receipt` CIDs in on the `Success` branch. As a Task may require subdelegation, the OPTIONAL `rec` field MAY be used to include recursive `Receipt`s.
+
+### 4.1.3 Metadata Fields
+
+The metadata field MAY be omitted or used to contain additional data about the receipt. This field MAY be used for tags, commentary, trace information, and so on.
+
+## 4.2 IPLD Schema
+
+``` ipldsch
+type SignedReceipt struct {
+  rec Receipt (rename "ucan/receipt")
+  sig VarSig
+}
+
+type Receipt struct {
+  inv  &SignedInvocation
+  out  {String : Result}
+  meta optional Any
+} 
+
+type Result union {
+  | Success
+  | Failure
+}
+
+type Failure struct {
+  err   nullable String
+  trace Any
+}
+
+type Success struct {
+  val Any
```

Comment:
Q: would we benefit from any more richness here? The point of the memoization is ofc to access past results, but at the moment there's no standard place where those results would be

Reply:
Ah, tell me more!

My thinking in leaving this open is that an output for Docker may look quite different from Wasm. My assumption (which is possibly incorrect) is that the shape of data will vary by the kind of job being run. We would then define "if you run a Tensorflow pipeline, we expect the format to be returned in this format".

But if there's some common structure that would be helpful across all job types, I'm definitely interested in exploring them 💯

Reply:
Yeah fair enough – certainly the result of calling a WASM function might just be a return value, which is decidedly less complex than Docker.

In Bacalhau we have a general concept that "jobs produce data" and hence a generic place in our output spec that describes what data they produced as an IPFS DAG. E.g. on input the user passes:

```jsonc
"outputs": ["public-stuff", "sensitive-stuff"]
```

and in the returned result they get:

```jsonc
"outputs": {
     "public-stuff": "ipfs://Qmyd4a4",
     "sensitive-stuff": "s3://sensitive-data.buckets.amazonaws.com/myresult",
}
```

So I guess the general capability is "named outputs" – we expect the task to produce this set of outputs, and in the result here they are with the labels specified.

But indeed, maybe that is too specific and we'd be better off defining some task types first and seeing if a general results structure falls out of that work.

Reply:
@expede seeing this concept of named outputs brings me back to this idea of outputs that are "spaces" (more like object spaces really), and could make sense more generally (esp docker use-cases)

Reply:
@zeeshanlakhani can you expand on that? Inferring a lot here, but do you mean introducing a selector syntax to make plugging together Docker containers supported at this level of the spec?

Reply:
Let me think out loud here at (⌚👀) right before I go to bed 😅 

> So I guess the general capability is "named outputs" – we expect the task to produce this set of outputs, and in the result here they are with the labels specified. [...] But indeed, maybe that is too specific and we'd be better off defining some task types first and seeing if a general results structure falls out of that work.

I feel like this would live in the `task-docker` spec, because it's specific going to be specific to each type 🤔 A previous version of this spec included JSONPointer, but we binned it in favour of using (metered) Wasm tasks for data transformation, since presumably we'll have that in IPVM and Bacalhau. There's an infinite number of formats that we can receive data in.

The markdown that this thread references no longer exists, and now looks like this:

```json
{
  "ok": {
    "public-stuff": "ipfs://Qmyd4a4",
    "sensitive-stuff": "s3://sensitive-data.buckets.amazonaws.com/myresult"
  }
}
```

We could define expected output shapes in Tasks as described above:

```json
{
  "with": "https://exmaple.com",
  "do": "foo/bar",
  "inputs": {"a": 1},
  "outputs": [
     "sensitive-stuff",
     { 
       "public-stuff": {
         "deeplynested": [
           { 
             "branchA": {/* ... */},
             "branchB": {/* ... */}
           },
           { 
             "branchC": {/* ... */},
             "branchD": {/* ... */}
           },
         ]
       }
     }
  ]
}
```

But now is the invocation spec also presumably doing schema enforcement. Schemas can get quite elaborate.

The current reasoning (which could be wrong!) is that we can push parsing to Wasm... which granted is a big hammer, but content-addressable, extensible, & reusable. In this world, schemata can live in a separate spec, and be paired up with the relevant Wasm CID.

Reply:
> In this world, schemata can live in a separate spec, and be paired up with the relevant Wasm CID.

Which sounds like a really good pattern actually 🤔🤔🤔

Reply:
> @zeeshanlakhani can you expand on that? Inferring a lot here, but do you mean introducing a selector syntax to make plugging together Docker containers supported at this level of the spec?

Thinking aloud, I was thinking more toward the `task-docker` spec for sure, but it seems like outputs for docker containers should *maybe* specify an output type that always refers to either an ipfs cid, an object store URL like s3 or similar, or even a file? Ideally, docker has to write output somewhere that's a resource/URL/file or output nothing at all. But, I'm probably limiting it too much?

Reply:
> Ideally, docker has to write output somewhere that's a resource/URL/file or output nothing at all.

This is indeed what Bacalhau does – we specify a "publisher" (atm at a Task level) that describes where the output from the invocation goes (IPFS, Estuary, Web3.Storage, etc). 

So far in https://github.com/ucan-wg/invocation/discussions/5 I have made that opaque as an `Any` type – because those options feel quite network-specific to me. But perhaps a similar level of abstraction to the `with` where everything should be expressable as a CID/URI is desirable...?